File: JeMPI_Apps/JeMPI_Shared_Source/custom/CustomLibMPIMutations.java
package org.jembi.jempi.libmpi.dgraph;

import java.util.UUID;

import org.jembi.jempi.shared.models.CustomPatient;
import org.jembi.jempi.shared.utils.AppUtils;

class CustomLibMPIMutations {

   private CustomLibMPIMutations() {}

   static String createPatientTriple(final CustomPatient patient, final String sourceUID) {
      final String uuid = UUID.randomUUID().toString();
      return String.format(
         """
         _:%s  <Patient.source_id>                <%s>        .
         _:%s  <Patient.aux_id>                   %s          .
         _:%s  <Patient.given_name>               %s          .
         _:%s  <Patient.family_name>              %s          .
         _:%s  <Patient.gender>                   %s          .
         _:%s  <Patient.dob>                      %s          .
         _:%s  <Patient.city>                     %s          .
         _:%s  <Patient.phone_number>             %s          .
         _:%s  <Patient.national_id>              %s          .
         _:%s  <dgraph.type>                     "Patient"    .
         """,
         uuid, sourceUID,
         uuid, AppUtils.quotedValue(patient.auxId()),
         uuid, AppUtils.quotedValue(patient.givenName()),
         uuid, AppUtils.quotedValue(patient.familyName()),
         uuid, AppUtils.quotedValue(patient.gender()),
         uuid, AppUtils.quotedValue(patient.dob()),
         uuid, AppUtils.quotedValue(patient.city()),
         uuid, AppUtils.quotedValue(patient.phoneNumber()),
         uuid, AppUtils.quotedValue(patient.nationalId()),
         uuid);
   }

   static String createLinkedGoldenRecordTriple(final CustomPatient patient,
                                                final String patientUID,
                                                final String sourceUID,
                                                final float score) {
      final String uuid = UUID.randomUUID().toString();
      return String.format(
         """
         _:%s  <GoldenRecord.source_id>                     <%s>             .
         _:%s  <GoldenRecord.aux_id>                        %s               .
         _:%s  <GoldenRecord.given_name>                    %s               .
         _:%s  <GoldenRecord.family_name>                   %s               .
         _:%s  <GoldenRecord.gender>                        %s               .
         _:%s  <GoldenRecord.dob>                           %s               .
         _:%s  <GoldenRecord.city>                          %s               .
         _:%s  <GoldenRecord.phone_number>                  %s               .
         _:%s  <GoldenRecord.national_id>                   %s               .
         _:%s  <GoldenRecord.patients>                      <%s> (score=%f)  .
         _:%s  <dgraph.type>                                "GoldenRecord"   .
         """,
         uuid, sourceUID,
         uuid, AppUtils.quotedValue(patient.auxId()),
         uuid, AppUtils.quotedValue(patient.givenName()),
         uuid, AppUtils.quotedValue(patient.familyName()),
         uuid, AppUtils.quotedValue(patient.gender()),
         uuid, AppUtils.quotedValue(patient.dob()),
         uuid, AppUtils.quotedValue(patient.city()),
         uuid, AppUtils.quotedValue(patient.phoneNumber()),
         uuid, AppUtils.quotedValue(patient.nationalId()),
         uuid, patientUID, score,
         uuid);
   }
}

File: JeMPI_Apps/JeMPI_Configuration/src/main/scala/configuration/CustomLibMPIMutations.scala
package configuration

import java.io.{File, PrintWriter}

private object CustomLibMPIMutations {

  private val classLocation = "../JeMPI_LibMPI/src/main/java/org/jembi/jempi/libmpi/dgraph"
  private val custom_className = "CustomLibMPIMutations"
  private val packageText = "org.jembi.jempi.libmpi.dgraph"

  def generate(fields: Array[Field]): Unit =
    val classFile: String = classLocation + File.separator + custom_className + ".java"
    println("Creating " + classFile)
    val file: File = new File(classFile)
    val writer: PrintWriter = new PrintWriter(file)
    writer.println(
      s"""package $packageText;
         |
         |import org.jembi.jempi.shared.models.CustomDemographicData;
         |import org.jembi.jempi.shared.utils.AppUtils;
         |
         |import java.util.UUID;
         |
         |final class $custom_className {
         |
         |${" " * 3}private $custom_className() {
         |${" " * 3}}
         |
         |${" " * 3}static String createPatientTriple(
         |${" " * 9}final CustomDemographicData demographicData,
         |${" " * 9}final String sourceUID) {
         |${" " * 6}final String uuid = UUID.randomUUID().toString();
         |${" " * 6}return String.format(\"\"\"""".stripMargin)

    // createDocumentTriple
    writer.println(s"""${" " * 27}_:%s  <PatientRecord.source_id>${" " * 16}<%s>${" " * 8}.""".stripMargin)
    fields.zipWithIndex.foreach {
      case (field, _) =>
        val name = field.fieldName
        writer.println(s"""${" " * 27}_:%s  <PatientRecord.$name>${" " * (25 - name.length)}%s${" " * 10}.""".stripMargin)
    }
    writer.println(
      s"""${" " * 27}_:%s  <dgraph.type>                     \"PatientRecord\"    .
         |${" " * 27}\"\"\",""".stripMargin)
    writer.println(s"""${" " * 27}uuid, sourceUID,""".stripMargin)
    fields.zipWithIndex.foreach {
      case (field, _) =>
        val name = "demographicData." + field.fieldName
        writer.println(
          s"""${" " * 27}uuid, AppUtils.quotedValue(${Utils.snakeCaseToCamelCase(name)}()),""".stripMargin)
    }
    writer.println(
      s"""${" " * 27}uuid);
         |   }
         |""".stripMargin)

    // createLinkedGoldenRecordTriple
    writer.println(
      s"""   static String createLinkedGoldenRecordTriple(
         |         final CustomDemographicData demographicData,
         |         final String patientUID,
         |         final String sourceUID,
         |         final float score) {
         |      final String uuid = UUID.randomUUID().toString();
         |      return String.format(\"\"\"
         |                           _:%s  <GoldenRecord.source_id>                     <%s>             .""".stripMargin)
    fields.zipWithIndex.foreach {
      case (field, _) =>
        val name = field.fieldName
        writer.println(
          s"""${" " * 27}_:%s  <GoldenRecord.$name>${" " * (30 - name.length)}%s${" " * 15}.""".stripMargin)
    }
    writer.println(
      s"""${" " * 27}_:%s  <GoldenRecord.patients>                      <%s> (score=%f)  .
         |${" " * 27}_:%s  <dgraph.type>                                "GoldenRecord"   .
         |${" " * 27}\"\"\",
         |${" " * 27}uuid, sourceUID,""".stripMargin)
    fields.zipWithIndex.foreach {
      case (field, _) =>
        val name = "demographicData." + Utils.snakeCaseToCamelCase(field.fieldName)
        writer.println(s"""${" " * 27}uuid, AppUtils.quotedValue($name()),""".stripMargin)
    }
    writer.println(
      s"""${" " * 27}uuid, patientUID, score,
         |${" " * 27}uuid);
         |${" " * 3}}
         |}""".stripMargin)
    writer.flush()
    writer.close()
  end generate

}
